compiler/qsc_project/src/manifest.rs

```diff
@@ -0,0 +1,90 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+use crate::Error;
+use serde::Deserialize;
+use std::path::PathBuf;
+use std::{
+    env::current_dir,
+    fs::{self, DirEntry, FileType},
+};
+
+pub const MANIFEST_FILE_NAME: &str = "qsharp.json";
+
+/// A Q# manifest, used to describe project metadata.
+#[derive(Deserialize, Debug, Default)]
+pub struct Manifest {
+    pub author: Option<String>,
+    pub license: Option<String>,
+    #[serde(default)]
+    pub exclude_files: Vec<String>,
+}
+
+/// Describes the contents and location of a Q# manifest file.
+#[derive(Debug)]
+pub struct ManifestDescriptor {
+    pub manifest: Manifest,
+    pub manifest_dir: PathBuf,
+}
+
+impl ManifestDescriptor {
+    pub(crate) fn new(manifest: Manifest, manifest_dir: PathBuf) -> Self {
+        Self {
+            manifest,
+            manifest_dir,
+        }
+    }
+}
+
+impl Manifest {
+    /// Starting from the current directory, traverse ancestors until
+    /// a manifest is found.
+    /// Returns an error if there are any filesystem errors, or if
+    /// a manifest file exists but is the wrong format.
+    /// Returns `Ok(None)` if there is no file matching the manifest file
+    /// name.
+    pub fn load() -> std::result::Result<Option<ManifestDescriptor>, Error> {
+        let current_dir = current_dir()?;
+        Self::load_from_path(current_dir)
+    }
+
+    /// Given a [PathBuf], traverse [PathBuf::ancestors] until a Manifest is found.
+    /// Returns [None] if no manifest named [MANIFEST_FILE_NAME] is found.
+    /// Returns an error if a manifest is found, but is not parsable into the
+    /// expected format.
+    pub fn load_from_path(path: PathBuf) -> std::result::Result<Option<ManifestDescriptor>, Error> {
+        let ancestors = path.ancestors();
+        for ancestor in ancestors {
+            let listing = ancestor.read_dir()?;
+            for item in listing.into_iter().filter_map(only_valid_files) {
+                if item.file_name().to_str() == Some(MANIFEST_FILE_NAME) {
+                    let mut manifest_dir = item.path();
+                    // pop off the file name itself
+                    manifest_dir.pop();
+
+                    let manifest = fs::read_to_string(item.path())?;
```

Comment:
I'm going to go out on a limb here and guess that fs::read_to_string doesn't work in the browser or VS Code.

Reply:
But this is always compiled in and part of the public API, yet on wasm this results in calls to this module:

<img width="310" alt="image" src="https://github.com/microsoft/qsharp/assets/993909/f7a7e7f6-4ecd-423f-822b-cede7e8c051a">

..which simply errors.

We shouldn't expose APIs by default we know are just going to unconditiionally fail on our main platforms.